### PR TITLE
feat: Neuron initialization and rescaling from Variance Transfert

### PR DESCRIPTION
**Pull request by TheoRudkiewicz**
Adds variance-transfer (VT) weight rescaling and (V,V)/(Z,-Z) neuron pairing to the growing-module extension workflow, with a comprehensive test suite validating correctness and edge cases.

See https://theorudkiewicz.github.io/gromo/tech_notes.html for updated tech notes.

### CODE REVIEW

File: src/gromo/modules/growing_module.py [hunk 1]
```diff
@@ -2945,6 +2945,260 @@ def create_layer_out_extension(self, extension_size: int) -> None:
...
+        Parameters
+        ----------
+        rescaling : _KNOWN_RESCALING_STRATEGIES_TYPE | None
+            Rescaling strategy.  One of ``"none"``, ``"default_vt"``,
+            ``"vt_constraint_old_shape"``, ``"vt_constraint_new_shape"``.
+        neuron_pairing : _KNOWN_NEURON_PAIRINGS_TYPE | None
+            Neuron-pairing strategy that will be applied *after* rescaling.
+            Needed to compute the effective extension size (pairing doubles
+            the extension).  One of ``"none"``, ``"vv_z_negz"``.
+        extension_size : int | None
```
[the remark this answers is not in the document]
Comment (TheoRudkiewicz) [reply src/gromo/modules/growing_module.py]: Fixed.

File: tests/test_variance_transfer.py [hunk 2]
```diff
@@ -0,0 +1,1097 @@
...
+        # = alpha*beta*conv_path_before + beta*(1 - alpha)*b2
+        # For default_vt (alpha=1), this simplifies to alpha*beta*conv_path.
+        # For B/C strategies, the b2 term introduces a small offset.
+        expected = alpha * beta * conv_path_before + skip
+        self.assertAllClose(
+            y_after,
+            expected,
+            atol=0.05,
+            msg=f"rescaling={rescaling}, alpha={alpha:.4f}, beta={beta:.4f}",
+        )
```
[the remark this answers is not in the document]
Comment (TheoRudkiewicz) [reply tests/test_variance_transfer.py:504]: Fixed.

File: src/gromo/modules/growing_module.py [hunk 3]
```diff
@@ -3118,6 +3409,10 @@ def create_layer_extensions(
         if layer_to_init.bias is not None:
             init_fn(layer_to_init.bias, self.previous_module.bias, prev_fan_in)
 
+        # Step 4: Neuron pairing (after init)
+        if neuron_pairing is not None:
+            self.apply_neuron_pairing(neuron_pairing=neuron_pairing)
+
```
[the remark this answers is not in the document]
Comment (TheoRudkiewicz) [reply src/gromo/modules/growing_module.py]: I think it's not a problem since for now there is no plan to use pairing with eigenvalues extension used.

File: docs/source/tech_notes/initialization.rst [hunk 4]
```diff
@@ -0,0 +1,122 @@
+Extension initialization
+========================
+
+When a layer of neurons is grown, new incoming and outgoing connections are
+created via
+:meth:`~gromo.modules.growing_module.GrowingModule.create_layer_extensions`.
+The method accepts an ``output_extension_init`` and an ``input_extension_init``
+argument that control how the new weight tensors are filled.
+
+The execution order inside ``create_layer_extensions`` is:
```
Comment (alexdavey) [inline docs/source/tech_notes/initialization.rst]: It's unclear why this is here -- shouldn't this be specific to the `variance_transfer` initialization setting?

File: docs/source/tech_notes/variance_transfer.rst [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
+     - :math:`h_t / h_{t+1}` (< 1)
+     - :math:`1` (exact)
+     - small init deficit, exact after 1 step
+
+
+Implementation
+--------------
+
+The variance-transfer features are exposed through
+:meth:`~gromo.modules.growing_module.GrowingModule.create_layer_extensions` via
```
Comment (alexdavey) [inline docs/source/tech_notes/variance_transfer.rst:394]: Typo: `meth` -> `math`. From a search there appears to be six of these in the codebase (including some pre-existing typos).

File: src/gromo/containers/growing_block.py [hunk 6]
```diff
@@ -593,27 +596,93 @@ def create_layer_extensions(
...
-            size of the input extension to create, if None use extension_size
-        output_extension_init: str
-            Initialization method for the output extension. Possible values include
-            "copy_uniform", "kaiming", "zeros", or other supported initialization
-            strategies.
-        input_extension_init: str
-            Initialization method for the input extension. Possible values include
-            "copy_uniform", "kaiming", "zeros", or other supported initialization
-            strategies.
+        extension_size : int
```
Comment (alexdavey) [inline src/gromo/containers/growing_block.py]: [Minor] The new "French" colon placement is inconsistent with the rest of the docstrings.

File: src/gromo/modules/growing_module.py [hunk 7]
```diff
@@ -2945,6 +2945,273 @@ def create_layer_out_extension(self, extension_size: int) -> None:
...
+        neuron_pairing: _KNOWN_NEURON_PAIRINGS_TYPE | None = None,
+        extension_size: int | None = None,
+    ) -> None:
+        """Rescale existing weights in-place before extension concatenation.
+
+        Implements three variance-transfer strategies from [1]_:
+
+        * ``"default_vt"`` (Strategy A): beta = sqrt(fan_in_old / fan_in_new),
+          alpha = 1 (the previous layer input is not extended).
+        * ``"vt_constraint_old_shape"`` (Strategy B): alpha and beta chosen so
```
Comment (alexdavey) [inline src/gromo/modules/growing_module.py:3279]: Why would we want to enforce variance with respect to the previous `fan_in`? If we consider a sequence of growth steps, the behaviour we get is:

- `default_vt`: Variance `\propto 1/current_fan_in` (with no training).
- `vt_constraint_new_shape`: Variance `\propto 1/current_fan_in`.
- `vt_constraint_old_shape`: Variance `\propto 1/penultimate_fan_in`.

The behaviour of the last option `vt_constraint_old_shape` does not really make sense to me.

**Also:** Should we consider making the gain a parameter in the constraint strategies? So that we can use e.g. kaiming `sqrt(2)/fan_in` instead?

File: src/gromo/modules/growing_module.py [hunk 8]
```diff
@@ -2945,6 +2945,273 @@ def create_layer_out_extension(self, extension_size: int) -> None:
...
+        # --- Fan-in values ---
+        fan_in_self_old = self.get_fan_in_from_layer(self.layer)
+
+        # When in_neurons == 0 (e.g. hidden_channels=0), there are no
+        # existing weights to rescale — nothing to do.
+        if self.in_neurons == 0:
+            return
+
+        # receptive_field = k*k for Conv2d, 1 for Linear
+        receptive_field = fan_in_self_old // self.in_neurons
```
Comment (alexdavey) [inline src/gromo/modules/growing_module.py]: This seems inconsistent with the approach elsewhere where the fan_in is computed in a centralised manner using `get_fan_in_from_layer`.

File: src/gromo/modules/growing_module.py [hunk 9]
```diff
@@ -3015,53 +3282,86 @@ def create_layer_extensions(
...
-        layer has a bias.
+        Additional initialization methods can be added by registering them in
+        the local ``known_inits`` dictionary of this method.  Each
+        initialization callable is applied to the extension weight tensor and
+        to the extension bias tensor, if the layer has a bias.
 
         The callable must accept the following arguments:
 
-        tensor: torch.Tensor
+        tensor : torch.Tensor
```
Comment (alexdavey) [inline src/gromo/modules/growing_module.py]: Similarly, this colon spacing is inconsistent with the rest of the codebase.

File: src/gromo/modules/growing_module.py [hunk 10]
```diff
@@ -3015,53 +3282,86 @@ def create_layer_extensions(
...
-            size of the output extension to create, if None use extension_size
-        input_extension_size: int | None
-            size of the input extension to create, if None use extension_size
-        output_extension_init: str
-            Initialization method for the output extension. Must be one of the keys in
-            `known_inits` ("copy_uniform", "kaiming", "zeros"), default "copy_uniform".
-        input_extension_init: str
-            Initialization method for the input extension. Must be one of the keys in
-            `known_inits` ("copy_uniform", "kaiming", "zeros"), default "copy_uniform".
+        extension_size : int
```
Comment (alexdavey) [inline src/gromo/modules/growing_module.py]: Similarly, this colon spacing is inconsistent with the rest of the codebase.

File: src/gromo/modules/growing_module.py [hunk 11]
```diff
@@ -3015,53 +3282,86 @@ def create_layer_extensions(
...
+            the keys in ``known_inits`` (``"copy_uniform"``, ``"kaiming"``,
+            ``"zeros"``), default ``"copy_uniform"``.
+        input_extension_init : str
+            Initialisation method for the input extension.  Must be one of
+            the keys in ``known_inits`` (``"copy_uniform"``, ``"kaiming"``,
+            ``"zeros"``), default ``"copy_uniform"``.
+        neuron_pairing : _KNOWN_NEURON_PAIRINGS_TYPE | None
+            Neuron-pairing strategy applied after initialisation.
+            ``"none"`` (default) or ``"vv_z_negz"``.
+            /!/ Note that neuron pairing doubles the effective extension size, so the
```
Comment (alexdavey) [inline src/gromo/modules/growing_module.py]: I didn't quite follow the discussion about this. Can't we keep the `extension_size` argument equal to the actual extension size, and instead internally halve the extension size if using `vv_z_neg`? This will require checking that it's even, but otherwise it's a much less bug-prone interface.

File: src/gromo/modules/growing_module.py [hunk 12]
```diff
@@ -3118,6 +3430,13 @@ def create_layer_extensions(
         if layer_to_init.bias is not None:
             init_fn(layer_to_init.bias, self.previous_module.bias, prev_fan_in)
 
+        # Step 4: Neuron pairing (after init)
+        if neuron_pairing is not None:
```
Comment (alexdavey) [inline src/gromo/modules/growing_module.py]: We don't need the `none` guard if `neuron_pairing=None` is a noop?

File: tests/test_variance_transfer.py [hunk 13]
```diff
@@ -0,0 +1,1125 @@
...
+
+
+def _grow_block(
+    block: GrowingBlock,
+    dh: int,
+    rescaling=None,
+    neuron_pairing=None,
+    output_extension_init: str = "kaiming",
+    input_extension_init: str = "kaiming",
+    noise_ratio: float = 0.0,
```
Comment (alexdavey) [inline tests/test_variance_transfer.py:83]: From what I can see, the path where the `noise_ratio` is non-zero is never exercised.

File: tests/test_variance_transfer.py [hunk 14]
```diff
@@ -0,0 +1,1125 @@
...
+            block,
+            dh=dh,
+            rescaling=rescaling,
+            neuron_pairing="vv_z_negz",
+        )
+
+        with torch.no_grad():
+            y_after = block(x)
+
+        # conv_path_before = W2*W1*x + W2*b1 + b2
```
Comment (alexdavey) [inline tests/test_variance_transfer.py]: [TODO] Check this

File: tests/test_variance_transfer.py [hunk 15]
```diff
@@ -0,0 +1,1125 @@
...
+
+
+class TestWeightVarianceAfterRescaling(TorchTestCase):
+    """Tests 3-4: weight variance of existing weights after rescaling.
+
+    These tests verify that ``apply_rescaling`` achieves the target variance
+    on the existing weight tensors *before* extension concatenation.
+
+    After rescaling:
+    - Conv1 (previous layer): ``alpha^2 * V[W1_old] = 1 / fan_in_prev``
```
Comment (alexdavey) [inline tests/test_variance_transfer.py]: [TODO] Why no strategy A?

File: docs/source/tech_notes/variance_transfer.rst [hunk 16]
```diff
@@ -0,0 +1,418 @@
...
+
+   \beta^2\, \operatorname{Var}[W_2^{(t)}] &= \frac{1}{h_t\, k^2}
+   \implies \operatorname{Var}[y'] = 1
+
+At init :math:`\operatorname{Var}[y] = 1` (exact).
+After one gradient step:
+
+.. math::
+
+   \operatorname{Var}[y]_{\text{1st}} = 1 + \frac{2\Delta h}{h_{t+1}} > 1
```
Comment (alexdavey) [inline docs/source/tech_notes/variance_transfer.rst]: [TODO] Should this and below be $h_t$ rather than $h_{t+1}$?

File: tests/test_variance_transfer.py [hunk 17]
```diff
@@ -0,0 +1,1125 @@
...
+        self.assertAlmostEqual(
+            var_w2,
+            1.0 / fan_in_self_new,
+            places=5,
+            msg=f"V[W2]={var_w2:.6f}, expected {1.0 / fan_in_self_new:.6f}",
+        )
+
+
+class TestActivationVariance(TorchTestCase):
+    """Tests 5-6: conv-path activation variance at init for Strategy B and C.
```
Comment (alexdavey) [inline tests/test_variance_transfer.py:611]: These tests don't seem to converge. According to codex, it's because the exact variance calculation does not include the border effects in the convolution and the bias contribution. Increasing the spatial resolution to 64x64 and disabling the biases, I can reduce `delta` to `0.1`. It's hard to reduce it further, but I didn't investigate why.

File: tests/test_variance_transfer.py [hunk 18]
```diff
@@ -0,0 +1,1125 @@
...
+            noise_ratio=0,
+        )
+        ext_out = block.first_layer.extended_output_layer
+        if ext_out.bias is not None:
+            self.assertTrue(
+                torch.equal(ext_out.bias[:dh], ext_out.bias[dh:]),
+                "Output extension bias: first half should equal second half",
+            )
+
+    def test_pairing_apply_change_then_forward_with_sized_activation(self):
```
Comment (alexdavey) [inline tests/test_variance_transfer.py:772]: I don't understand why we have this test.

File: tests/test_variance_transfer.py [hunk 19]
```diff
@@ -0,0 +1,1125 @@
+"""Tests for variance-transfer rescaling and neuron pairing.
```
Comment (alexdavey) [inline tests/test_variance_transfer.py:1]: Are there any bias tests?

File: tests/test_variance_transfer.py [hunk 20]
```diff
@@ -0,0 +1,1125 @@
...
+            extension_size=4,
+            output_extension_init="kaiming",
+            input_extension_init="kaiming",
+            rescaling=rescaling,
+            neuron_pairing="vv_z_negz",
+        )
+
+        # Weights should be nearly unchanged (tolerance accounts for
+        # finite-sample deviation of Kaiming uniform from exact 1/fan_in)
+        self.assertAllClose(
```
Comment (alexdavey) [inline tests/test_variance_transfer.py:825]: This does not converge, since the Conv2d initialisation is the standard pytorch one and not Kaiming. Correcting for the `sqrt(3)` factor yields a convergent test. But probably it's best to just use actual Kaiming for all these tests.

**Review by stephane-rivaud — APPROVED** [review]
I did not preview the docs changes nor did I thoroughly investigate the test cases, but the functional changes look good. I approve this PR.

**Comment by TheoRudkiewicz** [comment]
Before merging, I need to check that initialisation scale is good even with pairing.

**Comment by stephane-rivaud** [comment]
> Before merging, I need to check that initialisation scale is good even with pairing.

In `apply_rescaling`, we *roughly* define:
- `effective_size = 2 * size if neuron_paring == vt_negz else size`
- `fan_in = old_fan_in + effective_size * feature_size`

The bottom of the line is that we actually consider a tensor of size $2n$ in the resulting calculations.
This could be removed if we essentially prevented $n$ from being odd because we are indeed adding 2*n neurons.
I really wonder how the paper actually implements it.

**Comment by TheoRudkiewicz** [comment]
> Before merging, I need to check that initialization scale is good even with pairing.

There was indeed a problem, I should have fixed it in [d2d0d41129562d8e45b330377d01eaa79a3f5fa1](https://github.com/growingnet/gromo/pull/237/changes/d2d0d41129562d8e45b330377d01eaa79a3f5fa1).

**Comment by TheoRudkiewicz** [comment]
> > Before merging, I need to check that initialisation scale is good even with pairing.
> 
> In `apply_rescaling`, we _roughly_ define:
> 
>     * `effective_size = 2 * size if neuron_paring == vt_negz else size`
> 
>     * `fan_in = old_fan_in + effective_size * feature_size`
> 
> 
> The bottom of the line is that we actually consider a tensor of size 2 n in the resulting calculations. This could be removed if we essentially prevented n from being odd because we are indeed adding 2*n neurons. I really wonder how the paper actually implements it.



> > Before merging, I need to check that initialisation scale is good even with pairing.
> 
> In `apply_rescaling`, we _roughly_ define:
> 
>     * `effective_size = 2 * size if neuron_paring == vt_negz else size`
> 
>     * `fan_in = old_fan_in + effective_size * feature_size`
> 
> 
> The bottom of the line is that we actually consider a tensor of size 2 n in the resulting calculations. This could be removed if we essentially prevented n from being odd because we are indeed adding 2*n neurons. I really wonder how the paper actually implements it.

Yes I chose to ask for  growth of $n$ and with neuron pairing get a growth of $2n$.

**Comment by TheoRudkiewicz** [comment]
Remark from @alexdavey : we should probably include the "gain" in the re-scaling (even if variance transfer does not do it ?).

**Review by alexdavey — COMMENTED** [review]
Thanks @TheoRudkiewicz, here are a few comments :)

File: docs/source/tech_notes/variance_transfer.rst (hunk 5 above)
Comment (TheoRudkiewicz) [reply docs/source/tech_notes/variance_transfer.rst:394]: This is not a typo. `meth` is short for `method`
https://devguide.python.org/documentation/markup/#roles

File: src/gromo/containers/growing_block.py (hunk 6 above)
Comment (TheoRudkiewicz) [reply src/gromo/containers/growing_block.py]: Yes. Fixed.

File: src/gromo/modules/growing_module.py (hunk 10 above)
Comment (TheoRudkiewicz) [reply src/gromo/modules/growing_module.py]: Yes. Fixed.

File: src/gromo/modules/growing_module.py (hunk 7 above)
Comment (TheoRudkiewicz) [reply src/gromo/modules/growing_module.py:3279]: Why ?
Short answer it's the one propose by the paper VT.
Long answer: https://theorudkiewicz.github.io/gromo/tech_notes/variance_transfer.html#part-3-combined-analysis

About the gain parameter: probably.

File: src/gromo/modules/growing_module.py (hunk 9 above)
Comment (TheoRudkiewicz) [reply src/gromo/modules/growing_module.py]: Fixed.

File: src/gromo/modules/growing_module.py (hunk 12 above)
Comment (TheoRudkiewicz) [reply src/gromo/modules/growing_module.py]: Yes.

File: src/gromo/modules/growing_module.py (hunk 8 above)
Comment (TheoRudkiewicz) [reply src/gromo/modules/growing_module.py]: Yes. I fixed even if it required changing `get_fan_in_from_layer`.

File: src/gromo/modules/growing_module.py (hunk 11 above)
Comment (TheoRudkiewicz) [reply src/gromo/modules/growing_module.py]: Yes I agree. In the end it was not so difficult to switch so I did the change.

File: tests/test_variance_transfer.py (hunk 18 above)
Comment (TheoRudkiewicz) [reply tests/test_variance_transfer.py:772]: It was to illustrate that the extension size was different in `create_layer_extensions` and in `apply_change` due to neuron pairing.
1. The test was failing for the wrong reasons.
2. Now it does not fail anymore as sizes are compatible.

File: tests/test_variance_transfer.py (hunk 20 above)
Comment (TheoRudkiewicz) [reply tests/test_variance_transfer.py:825]: Yes this is now fixed.

File: tests/test_variance_transfer.py (hunk 17 above)
Comment (TheoRudkiewicz) [reply tests/test_variance_transfer.py:611]: To avoid border effects I went for fc layers. For some reason it is quiet slow to converge. I left high margin but I managed to reduce it locally with bigger shapes.

File: tests/test_variance_transfer.py (hunk 19 above)
Comment (TheoRudkiewicz) [reply tests/test_variance_transfer.py:1]: Yes, it just that the second extension does not have a bias.

File: tests/test_variance_transfer.py (hunk 13 above)
Comment (TheoRudkiewicz) [reply tests/test_variance_transfer.py:83]: Fixed.

**Comment by TheoRudkiewicz** [comment]
> > > Before merging, I need to check that initialisation scale is good even with pairing.
> > 
> > 
> > In `apply_rescaling`, we _roughly_ define:
> > ```
> > * `effective_size = 2 * size if neuron_paring == vt_negz else size`
> > 
> > * `fan_in = old_fan_in + effective_size * feature_size`
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The bottom of the line is that we actually consider a tensor of size 2 n in the resulting calculations. This could be removed if we essentially prevented n from being odd because we are indeed adding 2*n neurons. I really wonder how the paper actually implements it.
> 
> > > Before merging, I need to check that initialisation scale is good even with pairing.
> > 
> > 
> > In `apply_rescaling`, we _roughly_ define:
> > ```
> > * `effective_size = 2 * size if neuron_paring == vt_negz else size`
> > 
> > * `fan_in = old_fan_in + effective_size * feature_size`
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The bottom of the line is that we actually consider a tensor of size 2 n in the resulting calculations. This could be removed if we essentially prevented n from being odd because we are indeed adding 2*n neurons. I really wonder how the paper actually implements it.
> 
> Yes I chose to ask for growth of n and with neuron pairing get a growth of 2 n .

In the end we add $n$ neuron if we ask for $n$ neurons even with pairing.

**Comment by TheoRudkiewicz** [comment]
> Remark from @alexdavey : we should probably include the "gain" in the re-scaling (even if variance transfer does not do it ?).

I leave this for future work as it should be integrated in many places like `create_layer_extensions` with `kaiming`, `apply_rescaling` and `normalize_optimal_updates` fallback.

**Review by alexdavey — APPROVED** [review]
I didn't have enough time to fully double-check the discussion regarding strategies B and C unfortunately, but what you say makes sense. Otherwise everything LGTM :)
